Add pre-processed header support.

helly25 opened the pull request:
This is a expanded version of https://github.com/hedronvision/bazel-compile-commands-extractor/pull/205

It adds more header file extensions.

This fixes: https://github.com/hedronvision/bazel-compile-commands-extractor/issues/199

Commit:
Add pre-processed header support.

This is a expanded version of https://github.com/hedronvision/bazel-compile-commands-extractor/pull/205

It adds more header file extensions.

mikael-s-persson commented:
I just tried to use this patch, and it failed on one of my dependencies (mcap library) that has some `.inl` headers. That is a fairly common header('ish) extension too.

I'd suggest adding `.inl` to the list, I guess to the C list.

Commit:
Add '.inl' as possible c source.

There is an issue with those as they are typically not meant as stand alone sources.
Nonetheless, the tool should just gracefully warn on failure.

helly25 commented:
> I just tried to use this patch, and it failed on one of my dependencies (mcap library) that has some `.inl` headers. That is a fairly common header('ish) extension too.
> 
> I'd suggest adding `.inl` to the list, I guess to the C list.

I have added '.inl' as possible c source. However, there is an issue with those as they are typically not meant as stand alone sources. Nonetheless, the tool should just gracefully warn on failure.

mikael-s-persson commented:
I've noticed a couple of remaining issues with this. Example of a short excerpt of output:

```
....
clang: error: no such file or directory: 'bazel-out/k8-opt-exec-ST-d57f47055a04/bin/external/abseil-cpp~/absl/container/_objs/flat_hash_map/flat_hash_map.h.processed'
clang: warning: argument unused during compilation: '-fsyntax-only' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-fsyntax-only' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-fsyntax-only' [-Wunused-command-line-argument]
clang: error: no such file or directory: 'bazel-out/k8-opt-exec-ST-d57f47055a04/bin/external/abseil-cpp~/absl/time/_objs/time/civil_time.h.processed'
clang: error: no such file or directory: 'bazel-out/k8-opt-exec-ST-d57f47055a04/bin/external/abseil-cpp~/absl/time/_objs/time/clock.h.processed'
clang: error: no such file or directory: 'bazel-out/k8-opt-exec-ST-d57f47055a04/bin/external/abseil-cpp~/absl/time/_objs/time/time.h.processed'
clang: error: no such file or directory: 'bazel-out/k8-dbg/bin/external/abseil-cpp~/absl/container/_objs/flat_hash_map/flat_hash_map.h.processed'
....
```

The first issue is that the "-fsyntax-only" argument should be filtered out of the compilation arguments. I'm not familiar enough to know where exactly to do that but it seems like that should be trivial. This issue is not making it fail, but just creates a lot of noise in the output.

The second issue appears to me to be some sort of a race-condition on the ".processed" files. When re-running the generation again, the problem goes away, i.e., maybe the dependency on those output files is not getting properly registered/declared so that a more-or-less fresh run causes commands that expect those files to exist to be run before those files have been generated / sym-linked. Again, I don't know exactly how to fix that.

Commit:
Update patch to include `.ipp` and use configs instead of repeating them.
Ignore MODULE.bazel.lock

apesternikov commented:
I am confirming that this PR fixed my problem with abseil.
Thank you @helly25 !
Any plans on merging it?

helly25 commented:
> I am confirming that this PR fixed my problem with abseil. Thank you @helly25 ! Any plans on merging it?

Like others I am currently running on my own clone. The merge depends on Chris who is still in swallowed-up-by-google mode. When I asked him directly some time ago, he promised he would get to it once all legal issues are cleared. Sadly Google fired every single person I know who could have solved that. Until then we can only hope he enjoys his time and eventually gets to this.

helly25 commented:
FYI: As part of a viability sweep on the fork at https://github.com/helly25/bazel-compile-commands-extractor (upstream has been quiet for ~2 years), this PR has been rebased onto current fork main and merged as helly25/bazel-compile-commands-extractor#2.
